extensions/amp-subscriptions/0.1/local-subscription-platform-base.js

```diff
@@ -0,0 +1,192 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Actions} from './actions';
+import {
+  LocalSubscriptionPlatformRenderer,
+} from './local-subscription-platform-renderer';
+import {UrlBuilder} from './url-builder';
+import {closestAncestorElementBySelector} from '../../../src/dom';
+import {dev, userAssert} from '../../../src/log';
+
+
+/**
+ * This implements the methods to interact with various subscription platforms.
+ *
+ * @implements {./subscription-platform.SubscriptionPlatform}
+ */
+export class LocalSubscriptionBasePlatform {
+
+  /**
+   * @param {!../../../src/service/ampdoc-impl.AmpDoc} ampdoc
+   * @param {!JsonObject} platformConfig
+   * @param {!./service-adapter.ServiceAdapter} serviceAdapter
+   */
+  constructor(ampdoc, platformConfig, serviceAdapter) {
+    /** @protected @const */
+    this.ampdoc_ = ampdoc;
+
+    /** @private @const */
+    this.rootNode_ = ampdoc.getRootNode();
+
+    /** @protected {!JsonObject} */
+    this.serviceConfig_ = platformConfig;
+
+    /** @protected @const {!./service-adapter.ServiceAdapter} */
+    this.serviceAdapter_ = serviceAdapter;
+
+    /** @protected @const {!UrlBuilder} */
+    this.urlBuilder_ = new UrlBuilder(
+        this.ampdoc_,
+        this.serviceAdapter_.getReaderId('local'));
+
+    /** @protected @const {!./analytics.SubscriptionAnalytics} */
+    this.subscriptionAnalytics_ = serviceAdapter.getAnalytics();
+
+    userAssert(this.serviceConfig_['actions'],
+        'Actions have not been defined in the service config');
+
+    /** @private @const {!Actions} */
+    this.actions_ = new Actions(
+        this.ampdoc_, this.urlBuilder_,
+        this.subscriptionAnalytics_,
+        this.validateActionMap(this.serviceConfig_['actions'])
+    );
+
+    /** @private @const {!LocalSubscriptionPlatformRenderer}*/
+    this.renderer_ = new LocalSubscriptionPlatformRenderer(this.ampdoc_,
+        serviceAdapter.getDialog(), this.serviceAdapter_);
+  }
+
+  /**
+   * @override
+   */
+  getServiceId() {
+    return 'local';
+  }
+
+  /**
+   * Validates the action map
+   * @param {!JsonObject<string, string>} actionMap
+   * @return {!JsonObject<string, string>}
+   */
+  validateActionMap(actionMap) {
+    userAssert(actionMap['login'],
+        'Action "login" is not present in action map');
+    userAssert(actionMap['subscribe'],
+        'Action "subscribe" is not present in action map');
+    return actionMap;
+  }
+
+  /**
+   * Add event listener for the subscriptions action
+   * @protected
+   */
+  initializeListeners_() {
```

Comment:
Why not just keep this private and call from the base constructor? What's the reason to leaving it up to the subclasses to call this method?

Reply:
I didn't want to call it until after the subclass has initialized to avoid any possible race conditions.  I don't think there are any but I was being defensive.